src/main/scala/shine/OpenCL/CodeGeneration/InsertMemoryBarriers.scala

```diff
@@ -0,0 +1,208 @@
+package shine.OpenCL.CodeGeneration
+
+import shine.DPIA.ImperativePrimitives._
+import shine.DPIA.Phrases.{VisitAndRebuild, _}
+import shine.DPIA.Types._
+import shine.DPIA._
+import shine.OpenCL.ImperativePrimitives._
+import shine._
+
+import scala.collection.mutable
+
+object InsertMemoryBarriers {
+  def apply(p: Phrase[CommType]): Phrase[CommType] = {
+    val (p2, m) = visitNewMetadata(p, Map())
+    p2
+  }
+
+  private case class Metadata(// reads from outer scope allocations
+                              reads: mutable.Map[Identifier[_ <: PhraseType], AddressSpace],
+                              // work-group parallel writes to outer scope allocations
+                              wg_writes: mutable.Map[Identifier[_ <: PhraseType], AddressSpace])
+
+  private def visitNewMetadata(p: Phrase[CommType],
+                               // allocations in the current scope
+                               allocs: Map[Identifier[_ <: PhraseType], AddressSpace]
+                              ): (Phrase[CommType], Metadata) = {
+    val meta = Metadata(mutable.Map(), mutable.Map())
+    val p2 = VisitAndRebuild(p, Visitor(allocs, meta))
+    (p2, meta)
+  }
+
+  private def visitLoopBody(
+    p: Phrase[CommType],
+    allocs: Map[Identifier[_ <: PhraseType], AddressSpace],
+    metadata: Metadata,
+    outer_wg_writes: mutable.Map[Identifier[_ <: PhraseType], AddressSpace] = mutable.Map()
```

Comment:
Does `outer_wg_writes` need to be mutable here?

Reply:
It does not, I used a mutable map because that is what callers happen to have. I can change it to an immutable map if you want (not sure if the data structure conversion is free or not, probably not).

Reply:
Ok. I would have preferred to keep the interface of this function clean and make the map immutable, but I made a small performance test and it seems that the conversion from mutable to immutable is not cheap (I don't really get why. Can they not just keep the data unchanged and change the interface over it?).

In my super simple test, filling 1 000 000 values in a mutable Map takes about 140ms and converting it to an immutable Map takes about 530ms.

So I suggest you keep it.

Reply:
I assume that they have a different data structure implementation optimised for immutable operations than the one optimised for mutable operations, so they just rebuild the entire map on conversion.